pkg/build/buildkit/runner.go

```diff
@@ -0,0 +1,167 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package buildkit
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/moby/buildkit/client"
+	"github.com/okteto/okteto/pkg/analytics"
+	"github.com/okteto/okteto/pkg/env"
+	"github.com/okteto/okteto/pkg/log/io"
+)
+
+const (
+	// defaultMaxAttempts is the default number of attempts to try to do a build
+	defaultMaxAttempts = 3
+
+	// MaxRetriesForBuildkitTransientErrorsEnvVar is the environment variable to set the number of retries to wait for buildkit to be available
+	MaxRetriesForBuildkitTransientErrorsEnvVar = "OKTETO_MAX_RETRIES_FOR_BUILDKIT_TRANSIENT_ERRORS"
+)
+
+type registryImageChecker interface {
+	GetImageTagWithDigest(tag string) (string, error)
+}
+
+type buildkitWaiterInterface interface {
+	WaitUntilIsUp(ctx context.Context) error
+}
+
+type buildkitClientFactory interface {
+	GetBuildkitClient(ctx context.Context) (*client.Client, error)
+}
+
+type runnerMetadata struct {
+	attempts int
+	// TODO: use this to track the time it takes to solve a build instead of the full build time
+	solveTime time.Duration
+}
+
+// Runner runs a build using buildkit
+type Runner struct {
+	clientFactory                      buildkitClientFactory
+	waiter                             buildkitWaiterInterface
+	solveBuild                         SolveBuildFn
+	registry                           registryImageChecker
+	logger                             *io.Controller
+	metadata                           *runnerMetadata
+	maxAttemptsBuildkitTransientErrors int
+}
+
+// SolveBuildFn is a function that solves a build
+type SolveBuildFn func(ctx context.Context, c *client.Client, opt *client.SolveOpt, progress string, ioCtrl *io.Controller) error
+
+// NewBuildkitRunner creates a new buildkit runner
+func NewBuildkitRunner(clientFactory buildkitClientFactory, waiter buildkitWaiterInterface, registry registryImageChecker, solver SolveBuildFn, logger *io.Controller) *Runner {
+	return &Runner{
+		clientFactory:                      clientFactory,
+		waiter:                             waiter,
+		solveBuild:                         solver,
+		maxAttemptsBuildkitTransientErrors: env.LoadIntOrDefault(MaxRetriesForBuildkitTransientErrorsEnvVar, defaultMaxAttempts),
+		logger:                             logger,
+		registry:                           registry,
+		metadata:                           &runnerMetadata{},
+	}
+}
+
+// Run executes a build using buildkit
+func (r *Runner) Run(ctx context.Context, opt *client.SolveOpt, outputMode string) error {
+	tag := r.extractTagsFromOpt(opt)
+	attempts := 0
+	var solveTime time.Duration
+	defer func() {
+		r.metadata.attempts = attempts
+		r.metadata.solveTime = solveTime
+	}()
+	for {
+		attempts++
+		if attempts > 1 {
+			r.logger.Logger().Infof("retrying build, attempt %d", attempts)
+			r.logger.Out().Warning("Buildkit connection failure. Retrying...")
```

Comment:
In the [other PR](https://github.com/okteto/okteto/pull/4528#pullrequestreview-2360853208), we mentioned using always the same name. Shall we change this to adhere?

```suggestion
			r.logger.Out().Warning("BuildKit service connection failure. Retrying...")
```